underfs/obs/src/main/java/alluxio/underfs/oss/OBSOutputStream.java

```diff
@@ -0,0 +1,164 @@
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs.oss;
+
+import alluxio.util.io.PathUtils;
+
+import com.google.common.base.Preconditions;
+import com.obs.services.ObsClient;
+import com.obs.services.exception.ObsException;
+import com.obs.services.model.ObjectMetadata;
+import org.apache.commons.codec.binary.Base64;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.security.DigestOutputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A stream for writing a file into OBS. The data will be persisted to a temporary directory on the
+ * local disk and copied as a complete file when the {@link #close()} method is called.
+ */
+@NotThreadSafe
+public final class OBSOutputStream extends OutputStream {
+  private static final Logger LOG = LoggerFactory.getLogger(OBSOutputStream.class);
+
+  /** Bucket name of the Huawei OBS bucket. */
+  private final String mBucketName;
+  /** Key of the file when it is uploaded to OBS. */
+  private final String mKey;
+  /** The local file that will be uploaded when the stream is closed. */
+  private final File mFile;
+  /** The OBS client. */
+  private final ObsClient mObsClient;
+
+  /** The outputstream to a local file where the file will be buffered until closed. */
+  private OutputStream mLocalOutputStream;
+  /** The MD5 hash of the file. */
+  private MessageDigest mHash;
+
+  /** Flag to indicate this stream has been closed, to ensure close is only done once. */
+  private AtomicBoolean mClosed = new AtomicBoolean(false);
+
+  /**
+   * Creates a name instance of {@link OBSOutputStream}.
+   *
+   * @param bucketName the name of the bucket
+   * @param key the key of the file
+   * @param client the OBS client
+   */
+  public OBSOutputStream(String bucketName, String key, ObsClient client) throws IOException {
+    Preconditions.checkArgument(bucketName != null && !bucketName.isEmpty(),
+        "Bucket name must not be null or empty.");
+    Preconditions.checkArgument(key != null && !key.isEmpty(),
+        "OBS object key must not be null or empty.");
+    Preconditions.checkArgument(client != null, "ObsClient must not be null.");
+    mBucketName = bucketName;
+    mKey = key;
+    mObsClient = client;
+
+    mFile = new File(PathUtils.concatPath("/tmp", UUID.randomUUID()));
```

Comment:
do we have a requirement that '/tmp' is accessible by Alluxio servers?. if yes, please call it out in the docs

Reply:
replaced this line with File.createTempFile